File: 2023/01.py
import re

with open("input") as input:
    lines = input.readlines()

    # part one
    print(sum(int((d := re.findall(r"\d", line))[0] + d[-1]) for line in lines))

    # part two
    D = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

    def to_digit(s):
        return str(D.index(s) + 1) if s in D else s

    total = 0
    for line in lines:
        # A few lines have overlapping words, e.g. 'eightwo',
        # that's what the positive lookahead '?=' is for
        d = [*map(to_digit, re.findall(rf"(?=({r'|'.join(D)}|\d))", line))]
        total += int(d[0] + d[-1])
    print(total)

File: 2023/01.rs
const INPUT: &str = include_str!("../input");

fn main() {
    let calibration_value: u32 = INPUT
        .lines()
        .map(|line| {
            line.chars()
                .find(|c| c.is_ascii_digit())
                .unwrap()
                .to_digit(10)
                .unwrap()
                * 10
                + line
                    .chars()
                    .rev()
                    .find(|c| c.is_ascii_digit())
                    .unwrap()
                    .to_digit(10)
                    .unwrap()
        })
        .sum();
    println!("{calibration_value}");
}

fn find_digit() {
    
}
